hw/dv/dpi/usbdpi/usbdpi.c

```diff
@@ -552,8 +554,11 @@ char usbdpi_host_to_device(void *ctx_void, svBitVecVal *d2p_data) {
   switch (ctx->state) {
     case ST_IDLE:
       if ((simctrl && simctrl->TracingEnabled() && (ctx->frame == 20)) ||
-          (ctx->frame > 2000))
-        finish();
+          (ctx->frame > 50)) {
+        printf("USB: usbdpi done, frame: %d, success: %d, state: %d\n",
+               ctx->frame, ctx->setBaudSuccess, ctx->state);
+        usbdpi_finish(ctx->setBaudSuccess);
```

Comment:
I chose this, since it is a good indication for usbdev working correctly. This did not work correcty with the buggy 2p macro. I guess I'm understanding this simulation success wrong. You use it as an indication, that the simulation worked and not that the DUT worked, right?

Reply:
I suspect ultimately the whole "host" side in this dpi needs a bit of a rethink. It is already fragile to timing, especially in things like when it looks for and sends Ack returns. In its current state it provides a first level sanity check and enough to get tracing when things go wrong.

I always wondered about doing a connector to the usbip interface (which would allow the simulation to be connected to by a linux host), but instead moved to testing on the FPGA.

Reply:
The simulation return code is not meant to catch internal errors, but to pass the return code of a simulation, e.g. a test failed or succeeded. If you'd like to catch internal problems, please use assertions or similar mechanisms.

Reply:
> but to pass the return code of a simulation, e.g. a test failed or succeeded.
Then this seems to be a misunderstanding. `ctx->setBaudSuccess = 1` corresponds to successful communication with the device. It is a (relatively weak) check for the correct functionality of the USB device (DUT).

Reply:
Thanks for adding a comment to the corresponding section. To me, this does not look perfect, but reasonable and sufficient for now. I am not sure I get Philipp's suggestions right, but I would prefer not to add a call to `uartdpi_finish()` from `usbdpi.c`.